LSP final candidate

norcalli opened the pull request:
There are a few outstanding questions left, like:

- What interface should we expose to the users?
- How should we handle protocol extensions/custom protocols?
For example, clangd supports passing `offsetEncoding` in capabilities, but this is a protocol extension, so the code must be done in lua if someone wants to pass it (this is optional, however. utf-16 works just fine).
```lua
vim.lsp.add_filetype_config {
  name = "clangd";
  filetype = {"c", "cpp"};
  cmd = "clangd -background-index";
  capabilities = {
    offsetEncoding = {"utf-8", "utf-16"};
  };
  on_init = vim.schedule_wrap(function(client, result)
    local offset_encoding = result.offsetEncoding
    if offset_encoding then
      print(offset_encoding)
      client.offset_encoding = offset_encoding
    end
  end)
}
```
- When should I schedule_wrap things for the user? Is there a downside to always doing it? It seems like sometimes, the error reporting can suffer a little bit due to schedule_wrap.

Current viml example interface:

```vim
call lsp#set_log_level("debug")

call lsp#add_filetype_config({
  \ 'name': 'lua',
  \ 'filetype': ['lua'],
  \ 'cmd': ['/home/ashkan/works/3rd/lua-language-server/run.sh'],
  \ })

call lsp#add_filetype_config({
  \ 'name': 'js',
  \ 'filetype': ['javascript'],
  \ 'cmd': [$JAVASCRIPT_LANGUAGE_SERVER_DIRECTORY.'/lib/language-server-stdio.js'],
  \ })

call lsp#add_filetype_config({
  \ 'name': 'lua2',
  \ 'filetype': 'lua',
  \ 'cmd': './run.sh',
  \ 'cmd_cwd': '/home/ashkan/works/3rd/lua-language-server/',
  \ })
```

Each server must have a unique name to identify them. There's a way to define configs from `viml`, but `lua` may be necessary for advanced configuations.

If one didn't want to use `lsp.add_config` and wanted more granular control than `filetype`, they could use `lsp.start_client` and then `lsp.attach_to_buffer`.

I ended up rewriting about 90% of [the existing PR](https://github.com/neovim/neovim/pull/10222). I tried to reduce the modules further, but I think this is as good as it gets. It could still be done, however, but putting `protocol` into `util` or `lsp`.

tjdevries committed:
Rename neovim to nvim

tjdevries committed:
Add line ending under certain circumstances

tjdevries committed:
Work on cleaning up some documentation

tjdevries committed:
Make omni-complete better

tjdevries committed:
Make location list or quickfix list available in publishDiagnostics

KillTheMule committed:
Add fake lsp to test the lsp pr

KillTheMule committed:
Add first test for textDocument/hover

KillTheMule committed:
Support setting responses, fix the test

KillTheMule committed:
Don't put a final newline into hover results

KillTheMule committed:
Start tests with malformed responses

h-michael committed:
Separate Request Message and Notification Message

Language Server Protocol defines two message types which are Request Message and Notification Message from client.
Now, built-in lsp client is implemented only Request Message.
So I impletent Notification Message.

h-michael committed:
Fix config_callbacks_spec

h-michael committed:
Remove pre and post autocmd hook

I want to let lsp lib simplify at first.
So I remove pre and post autocmd hook until needed them.

h-michael committed:
Remove optional callback from notify

h-michael committed:
Fix function keyword

h-michael committed:
Don't load default callback when load lsp modules

h-michael committed:
test: fix config_callbacks_spec

h-michael committed:
test: fix structures_spec

h-michael committed:
Divide into core callback implementation and builtin callback implementation

h-michael committed:
Remove default property from CallbackObject

Now, we provide builtin callbacks instead of default callbacks.
So I remove default property from CallbakcObject

h-michael committed:
Refactor builtin callbacks

h-michael committed:
Fix luacheck warnings

h-michael committed:
runtime: fix is_empty function in nvim.util

`table ~= {}` always returns false.
And I think is_empty function should be for table, so if is_empty receive another type instead of table, it throws error.
I think `next(table)` is better way that check whether table is empty or not.

h-michael committed:
runtime: get the callbacks with a specific filetype, return only filetype callbacks.

h-michael committed:
runtime: move set builtin callbacks functions to config callbacks module

h-michael committed:
runtime: refactor lsp api and lsp config modules

h-michael committed:
test: Add lsp callbacks module test

h-michael committed:
runtime: Remove autocmds module

I think client complex more than necessary.
So I remove autocmds module until we really need.

h-michael committed:
runtime: Move nvim.util.table.is_empty to vim.shared

h-michael committed:
runtime: Use error instead of returning nil

h-michael committed:
runtime: Remove unused util functions

h-michael committed:
runtime: Remove client property from plugin module

h-michael committed:
runtime: Use vim.shared.split instead of nvim.util.split

h-michael committed:
runtime: Rename server.lua to server_config.lua

h-michael committed:
runtime: Add bufnr to client request argument.

If we receive the language server response when our currnent buffer is different from we posted request to the languager server, we can't know that we should apply to which buffer.
So I add bufnr to request method argument.
If there is no bufnr argument parameter, the client use current bufnr as argument.

h-michael committed:
runtime: Rename unnecessary statements

h-michael committed:
runtime: Fix checking nil

h-michael committed:
runtime: Chnage request method arguments order

h-michael committed:
runtime: Update lsp.config modules and the lsp help doc

h-michael committed:
runtime: Add error handling to server_config.lua

h-michael committed:
runtime: Rename api.lua properties

h-michael committed:
runtime: Call only lsp.api module from viml

h-michael committed:
runtime: add get_text_document_params function

h-michael committed:
runtime: add lsp util to api module

h-michael committed:
runtime: fixup client.lua's nil case

h-michael committed:
runtime: Use stdpath('data') for log directory

h-michael committed:
doc: update lsp document

h-michael committed:
runtime: rename local variable

h-michael committed:
runtime: Refactor plugin interface

h-michael committed:
runtime: don't use metatable method __len

h-michael committed:
runtime: Fix wrong arguments in client#request

h-michael committed:
runtime: Fix completion modules

h-michael committed:
runtime: Change log format

h-michael committed:
runtime: Refactor job handling

Translate job handling and starting client codes to Lua from Viml.

h-michael committed:
test: Fix some test cases

h-michael committed:
runtime: Fix hover callback

h-michael committed:
runtime: Chnange lsp.api to vim.lsp

h-michael committed:
Update runtime/doc/lsp.txt

Co-Authored-By: Daniel Hahler <github@thequod.de>

h-michael committed:
runtime: Fix import lsp module

h-michael committed:
runtime: Change log module

Add only client request and notification log and only server response and notification log for debug

h-michael committed:
runtime: Fix initialize client

h-michael committed:
runtime: Move plugin module implementation to vim.lsp module

h-michael committed:
runtime: Enable to set language client log level

h-michael committed:
runtime: Remove validations which is called before request

h-michael committed:
runtime: Use luv instead of nvim job for handling client

h-michael committed:
runtime: Fix completion implementation

h-michael committed:
runtime: Change request and request_async arguments order

h-michael committed:
runtime: Remove lsp-request Vim script wrappers

h-michael committed:
runtime: Implement textDocument/hover built-in callback

h-michael committed:
runtime: Fix floating window vertical position

h-michael committed:
runtime: Add URI module for transforming filepath to URI

h-michael committed:
Update runtime/lua/uri.lua

Co-Authored-By: Marco Hinz <mh.codebro+github@gmail.com>

Update runtime/lua/uri.lua

Co-Authored-By: Marco Hinz <mh.codebro+github@gmail.com>

h-michael committed:
runtime: Change output log path

h-michael committed:
runtime: fix uri module

h-michael committed:
test: Add uri module test

h-michael committed:
runtime: Fix textDocument/hover callback

h-michael committed:
runtime: Implement textDocument/signatureHelp callback

h-michael committed:
runtime: Remove start language server autocmd after starting

h-michael committed:
Update runtime/autoload/lsp/text_document.vim

Co-Authored-By: Marco Hinz <mh.codebro+github@gmail.com>

h-michael committed:
Apply suggestions from code review

Co-Authored-By: Justin M. Keyes <justinkz@gmail.com>

h-michael committed:
runtime: Use once autocmd

h-michael committed:
runtime: Remove unused modules and functions

h-michael committed:
runtime: Move uri.lua to vim/uri.lua and export to vim.lua

h-michael committed:
runtime: Use vim.* instead of require('vim.shared')

h-michael committed:
runtime: Move lsp/*.lua to vim/lsp/*.lua

h-michael committed:
runtime: Remove and move nvim/*.lua modules and methods

h-michael committed:
runtime: Add not implemented log

h-michael committed:
runtime: Fix autoload lsp.vim

h-michael committed:
runtime: Megrge structures module into protocol module

h-michael committed:
runtime: Megrge json module into util module

h-michael committed:
runtime: Clean up config module

h-michael committed:
runtime: Fix wrong requiring logger module

h-michael committed:
runtime: Fix textDocument/hover callback

If string contains new lines, we should split it with '\n'.

h-michael committed:
runtime: Fix wrong replacement

h-michael committed:
runtime: Implement textDocument/definition callback

h-michael committed:
runtime: Enable to set language client capabilities

h-michael committed:
runtime: Add client_info API for debug

h-michael committed:
runtime: Use vim.split instead of nvim_call_function('split', ...)

h-michael committed:
runtime: Rename fname_to_uri to uri_from_fname

h-michael committed:
runtime: Fix doc/lsp.txt format

h-michael committed:
test: Move and fix lua/util_spec.lua to lua/shared_spec.lua

h-michael committed:
test: Move and fix lsp/setup_spec.lua to lsp/server_config_spec.lua

h-michael committed:
doc: Fix wrong description

h-michael committed:
runtime: Add status method for debug

h-michael committed:
runtime: Use uri method instead of hard coding

h-michael committed:
runtime: Change file name from camelcase to snakecase

h-michael committed:
runtime: Refactor textDocument/hover callback

h-michael committed:
runtime: Change callback module confusing method names

h-michael committed:
runtime: Print shutting down message

h-michael committed:
runtime, test: Fix message.lua

h-michael committed:
runtime, Remove cmd argument from lsp.start_client method

h-michael committed:
runtime, test: Fix lsp API and update lsp API spec

h-michael committed:
test: Fix protocol_spec.lua

h-michael committed:
test: Remove the incorrect test once

h-michael committed:
test: Fix server_config_spec.lua

h-michael committed:
runtime: Fix callback for error response

h-michael committed:
runtime: Add tbl_keys and tbl_values to shared.lua

h-michael committed:
runtime: Remove Enum util

h-michael committed:
test: Fix integration_spec.lua

h-michael committed:
runtime: Remove check module

I delete it because I think it is too strong for the client library.

h-michael committed:
runtime: Enable to set multiple language server for one filetype

h-michael committed:
runtime: When language client stop, it does not request shutdown to the language server

Since it takes time to complete the shutdown request, do not use it in default.

h-michael committed:
runtime: Update tagstack when jump to a symbol

h-michael committed:
runtime: Add textDocument/declaration, typeDefinition, implementation

h-michael committed:
runtime: Add uv.kill after uv.shutdown

h-michael committed:
test: Fix message_spec.lua

h-michael committed:
test: Use ./build/bin/nvim

h-michael committed:
test: Fix doc syntax mistakes

h-michael committed:
test: Stop lsp after examples finishing

h-michael committed:
runtime: Add textDocument/didOpen, didChange, didSave

h-michael committed:
runtime: Fix autocmd for text synchronization method

h-michael committed:
runtime: Return true in the text_document_did_change_handler for detaching if the client stopped

h-michael committed:
runtime: Add documentation to completion preview

h-michael committed:
test: Remove integration_spec

Fake lsp architecture with headless nvim does not work well and it was very difficult to debug it.
So I remove these specs. Instead, I'll  add more unit tests later.

h-michael committed:
runtime: Fix builtin_callbacks module export

h-michael committed:
runtime: Manage version each documents, and update each textDocument/didChange

h-michael committed:
runtime: Remove DefaultMap

h-michael committed:
runtime: Remove unused function from lsp.util module

h-michael committed:
test: Add lsp.util spec

h-michael committed:
runtime: Move completion handler module to text document handler module

h-michael committed:
runtime: Rename handle modules to handler

h-michael committed:
runtime: Move SignatureHelp and Hover response contents data converting process to text document handler module

h-michael committed:
test: Add lsp.util spec and lsp.handler spec

h-michael committed:
doc: Fixup!

h-michael committed:
runtime: Remove language_client_enable_shutdown_request option

h-michael committed:
runtime: Remove unnecessary code

h-michael committed:
runtime: Fix check_table method

h-michael committed:
runtime: Fix completion item handling

h-michael committed:
runtime: Use jobstart style argument for server_config and add the offset_encoding option

h-michael committed:
runtime: Fix check_table method

h-michael committed:
test: Fix server_config_spec.lua

h-michael committed:
runtime: Fix check_table method

h-michael committed:
runtime: Fix offset encoding setting handling

h-michael committed:
runtime: Rename add_server_config arguments

h-michael committed:
test: Fix protocol and add more protocol spec examples

h-michael committed:
runtime: Fill completion-item info with white space when documentation is blank

h-michael committed:
runtime: Refactor lsp client module

- Remove unused variables
- Remove unused function
- Add _ prefix to private variables and functions

h-michael committed:
runtime: Refactor lsp logger module

h-michael committed:
runtime: Use vim.inspect instead of vim.tbl_tostring

h-michael committed:
runtime: Fix setting multiple language server for one filetype

h-michael committed:
runtime: Add omnifunc

h-michael committed:
runtime: Refactor modules

h-michael committed:
doc: Update auto completion doc

h-michael committed:
runtime: Refactor client module

h-michael committed:
Remove callback argument from lsp.request method

h-michael committed:
runtime: Raise error when the given command is not executable

h-michael committed:
runtime: Fixup lsp.logger module

h-michael committed:
runtime: Fixup logger and printing messages

h-michael committed:
test: Move shared_spec.lua to utility_function_spec.lua

Commit:
Checkpoint.

Commit:
Fix markdown type. Trip stuff.

Commit:
Clean up uri encoding stuff.

Commit:
Remove message.lua and util.lua

Integrate them into their callsites.

Commit:
Finish client rework.

Commit:
Wooh. final refactoring.

rpc is now detached from lsp logic.

callback is likely unnecessary.

Commit:
Bugfixes

Commit:
The didChange is currently broken, so send the full buffer for now.

Commit:
Clean up encoding validation.

Commit:
Clean up and handle errors.

Commit:
Remove callbacks code.

Commit:
Clean up comments.

Commit:
Start capability resolution rewrite

Add vim.tbl_deep_merge to help

TODO: use the resolved_capabilities in the handler initializations to
conditionally send things to the server.

Commit:
Merge remote-tracking branch 'origin/master' into lsp-review

Commit:
Change logger to allow lazy eval.

Commit:
Start using resolved_capabilities.

Commit:
Improve capabilities detection

Commit:
Slight cleanup

Commit:
Switch to spaces instead of tabs.

Commit:
Switch to spaces instead of tabs pt 2

Commit:
Further cleanup.

Add env and cwd options.
Improve documentation.
Improve log statements.

Commit:
Clean up protocol.

Commit:
Bugfix and enable signatureHelp

Commit:
Clean up unsupported_method handling.

Commit:
Final cleanup.

Commit:
Last cleanup.

Commit:
Remove tests for now.

Commit:
Remove out of date docs.

bfredl commented on runtime/lua/vim/lsp.lua:
```diff
@@ -0,0 +1,791 @@
+      return nil, "TIMEOUT"
+    end
+    -- TODO this really needs to be further looked at.
+    nvim_command "sleep 10m"
```
we have `vim.fn.wait(condition)` but it only works for vimL callbacks. Though exposing the same in lua with lua cb would be straightforward.

bfredl replied:
Then there is the question, if this is called from non-main coroutine, should it automagically yield instead (lua-client does this), or a separate `lsp.buf_request_yield` might be clearer/simpler.

norcalli replied:
As long as the semantics and side effects are the same, I’d probably yield in a coroutine and wait otherwise. Fair point either way.

bfredl commented:
> For example, clangd supports passing offsetEncoding in capabilities, but this is a protocol extension, so the code must be done in lua if someone wants to pass it (this is optional, however. utf-16 works just fine).

I think this (and most things) work in vimL already, it just forwards the entire config dict.

> When should I schedule_wrap things for the user? Is there a downside to always doing it? It seems like sometimes, the error reporting can suffer a little bit due to schedule_wrap.

How is the error handling worse? I think the sane default for a callback parameter is to automatically schedule it, as you can't access the editor without it (expect for a few things like nvim_input), plus an optional flag for "fast" callback.

andreypopp commented on runtime/lua/vim/lsp.lua:
```diff
@@ -0,0 +1,791 @@
+  return client_request_ids, cancel_request
+end
+
+--- Send a request to a server, but don't wait for the response
```
"but don't wait for the response" doesn't describe this, does it?

andreypopp commented on runtime/lua/vim/lsp.lua:
```diff
@@ -0,0 +1,791 @@
+    assert(type(conf.cmd_env) == 'table', "conf.cmd_env must be a table")
+  end
+  if conf.cmd_cwd then
+    assert(type(conf.cmd_cwd) == 'string', "conf.cmd_cwd must be a string")
```
I think it's useful to make `cmd_cwd` a function which receives buffer as arg and returns a directory which will be used as a cwd for LSP server and treated as a project root. That way inside neovim we could have several instances of LSP servers running for the same filetype but for different project.

norcalli replied:
the `cmd_cwd` is strictly for launching the cmd. This is for cases where the LSP command may expect to be launched from a directory (like if it has relative file pointers). What you're talking about is the `rootUri`, which does raise a very good point which I didn't consider which is configuring multiple of the same LSP for different projects within a single instance of neovim.

The question is whether that is a common enough use-case to warrant a configuration option via `lsp.add_config`. `lsp.add_config` is just for making the common case of launching a single LSP for multiple filetypes easier. If you want a more unique use-case, `lsp.start_client & lsp.attach_to_buffer` is the recommended solution and it is still very easy to use. I think if the documentation explains that clearly and has good examples, then it should be fine.

Also, what you're talking about may be covered by the `workspaceFolder` part of LSP, where a single workspace can have multiple folders.

If you wouldn't mind, @andreypopp  it sounds like you have a specific use-case in mind, so I'd be interested in seeing if you and I could get your existing configuration working with this branch. That would be a good test case. My use-case has been generally more straightforward so far.

andreypopp reviewed (commented):
Looks good!

It's not clear to me if it's possible to have multiple LSP server instances of the same server/configuration running at the same time (say I have files open from different projects).

norcalli commented on runtime/lua/vim/lsp.lua:
```diff
@@ -0,0 +1,791 @@
+-- @param server_name [string] (optional)
+--
+-- @returns: The table of request id
+function lsp.buf_request_sync(bufnr, method, params, timeout_ms)
```
The result will be passed in the return value since it is synchronous. I forgot to change the docs on that. It's a table of `{[client_id] = request_result}`

h-michael replied:
oops, sorry. I mistaken it as 'reques_async'.

norcalli replied:
Yeah I purposely chose to make the `_sync` case special since it's more atypical and also potentially more faulty. The sleep logic is still something I want to figure out a better solution for.

norcalli commented on runtime/lua/vim/lsp.lua:
```diff
@@ -0,0 +1,791 @@
+--   cmd = string;
+--   cmd_args = table;
+--   cmd_cwd = string | nil;
+--   cmd_end = table | nil;
```
libuv expects an array of `k=v` such as `{ "PRODUCTION=false"; }`, but I wrote a helper function to allow you to specify like a map as well. So it accepts both forms:

```
{
  "PRODUCTION=true";
  "B=b";
  PORT = 123;
  A = "testing";
}
```

norcalli commented on runtime/lua/vim/lsp.lua:
```diff
@@ -0,0 +1,791 @@
+local builtin_default_server_callbacks = require 'vim.lsp.builtin_callbacks'
```
Unless I'm misunderstanding you, that's not exactly true since the `default_server_callbacks` passed in will be `tbl_extend("keep", default_server_callbacks, builtin_server_callbacks)`, which means that we only provide ones the user doesn't provide.

I do agree that they should be available, but I thought that they would be via the require. You mean to expose them on the `lsp` object itself, right?

h-michael replied:
I'm sorry, I overlooked that. 
Regarding the public method, I think that this is related to the future policy of the Lua modules of the Neovim project, but I think that it is better to make public APIs available without requiring them. 
This makes it easier for users to distinguish between public and private APIs.  Also, in this example, I thought it would be redundant to call 'require (' vim.lsp.builtin_callbacks ')' for something that can be called 'vim.lsp.builtin_callback'.

@bfredl What do uou think about this?

norcalli commented on runtime/lua/vim/lsp.lua:
```diff
@@ -0,0 +1,791 @@
+-- @param server_name [string] (optional)
+--
+-- @returns: success?, request_id, cancel_fn
+function lsp.buf_request(bufnr, method, params, callback)
```
I'm not sure, but I thought it would make it very clear what it was doing. It is operating on the buffer, and it kind of follows other naming schemes for neovim functions.

norcalli commented on runtime/lua/vim/lsp/rpc.lua:
```diff
@@ -0,0 +1,357 @@
+
+  local handle, pid
+  do
+    local function onexit(code, signal)
```
The function exposed to the user is `on_exit`, actually. I just used `onexit` internally because I don't like to be consistent :P

norcalli commented:
@bfredl what I meant was more complicated setups would require using callbacks for configuration, which, afaik, precludes vimL from being used for configuration in that case since I don't think you could pass a viml callback to lua and modify a lua object from viml (which may contain lua functions and other special types).

Commit:
Improve tests and comments.

Commit:
Oopsie

Commit:
Bugfix and improve docs and change conf signature.

Commit:
Refactor and bugfix.

- Change callbacks to have a uniform signature.
- Clean up validation.
- Clean up scheduling.

Commit:
Small refactor.

Commit:
Fix incremental text.

Commit:
Add copy_config.

Commit:
Clean up comments.

Commit:
Clean up.

Commit:
Refactor little.

Commit:
Refactor once more.

Commit:
Use vim.wait when it's available or a fallback

Commit:
Redo shutdown mechanism

- Handle clients that haven't been uninitialized yet.
- Remove timers/timeouts other than at exit.
- Add stop escalation to term.

Commit:
Merge remote-tracking branch 'origin/master' into lsp-review

h-michael commented on test/functional/lua/utility_functions_spec.lua:
```diff
@@ -302,6 +302,7 @@ describe('lua stdlib', function()
       pcall_err(exec_lua, [[return vim.pesc(2)]]))
   end)
 
+<<<<<<< HEAD
```
conflict marker

Commit:
Oopies daisy.

Commit:
Finalize the interface and refactor

- Refactor builtin callbacks and documentation.
- Rename add_config to add_filetype_config.

Commit:
Almost finish the docs.

Commit:
Mostly finish lsp docs.

Commit:
Add client.is_stopped()

KillTheMule commented on runtime/doc/lsp.txt:
```diff
@@ -0,0 +1,647 @@
+      A unique identifying string among all other servers configured with
+      |vim.lsp.add_filetype_config|.
+ 
+  Differences:~
```
Differences to what?

KillTheMule commented on runtime/doc/lsp.txt:
```diff
@@ -0,0 +1,647 @@
+      `{ "PRODUCTION=true"; "TEST=123"; PORT = 8080; HOST = "0.0.0.0"; }`
+ 
+  `capabilities`
+    A {table} which will be merged using |vim.deep_merge| with neovim's
```
Isn't `deep_merge` an implementation detail that should probably be skipped? Just `merged` seems fine here.

norcalli replied:
I want to make it clear how it will be merged since it's a deeply nested object.

KillTheMule commented on runtime/doc/lsp.txt:
```diff
@@ -0,0 +1,647 @@
+  You can also use `client.stop()` if you have access to the client.
+
+                                                     *vim.lsp.stop_all_client()*
+vim.lsp.stop_all_client([{force}])
```
Shouldn't this be named `stop_all_clients`?

Commit:
Fix lualint errors.

Commit:
Get rid of upper case things.

Commit:
Make some basic integration tests.

Improve documentation and some edge cases as revealed by writing the
integration tests.

Commit:
Change client facade to be more functional.

Commit:
Fix test failures.

Commit:
Increase timeout for init test.

Commit:
Only use tohex if it's available.

glacambre commented on runtime/lua/vim/uri.lua:
```diff
@@ -29,7 +28,12 @@ do
+    tohex = require'bit'.tohex
+  else
+    tohex = function(b) return string.format("%02x", b) end
+  end
```
I think this might not be necessary, if people chose to build neovim with another lua than luajit they're supposed to make the bitop package a dependency ( https://github.com/neovim/neovim/issues/11301#issuecomment-546668666 ).

norcalli replied:
Our tests don't seem to agree :P https://travis-ci.org/neovim/neovim/jobs/609437664?utm_medium=notification&utm_source=github_status the travis test doesn't have bitop available.

mcepl replied:
> Our tests don't seem to agree :P https://travis-ci.org/neovim/neovim/jobs/609437664?utm_medium=notification&utm_source=github_status the travis test doesn't have bitop available.

Are you sure? https://travis-ci.org/neovim/neovim/jobs/609437664#L4593

bfredl replied:
what is poorly setup? the test seems to indicate that this code is needed for now, just as claimed.

Commit:
This test is non-deterministic, so cover either case.

norcalli commented on runtime/lua/vim/lsp/builtin_callbacks.lua:
```diff
@@ -0,0 +1,301 @@
+
+local function split_lines(value)
+  return vim.split(value, '\n', true)
+end
```
Yeah, but just because something is repeated twice, doesn't mean it's necessarily helpful to make it a util. I personally try not to be too DRY.

bfredl commented:
I get `LSP[clangd]: Error INVALID_SERVER_JSON: "\"E5560: vimL function must not be called in a lua loop callback\""` with master-rebase. Somewhere `vim.fn` is used without `vim.schedule[_wrap]`, this needs to be changed.

bfredl commented:
This works for now:
```diff
diff --git a/runtime/lua/vim/lsp/rpc.lua b/runtime/lua/vim/lsp/rpc.lua
index ab44a2914..52ce44491 100644
--- a/runtime/lua/vim/lsp/rpc.lua
+++ b/runtime/lua/vim/lsp/rpc.lua
@@ -389,7 +389,7 @@ local function create_and_start_client(cmd, cmd_args, handlers, extra_spawn_para
       local headers, body = request_parser(chunk)
       -- If we successfully parsed, then handle the response.
       if headers then
-        handle_body(body)
+        vim.schedule(function() handle_body(body) end)
         -- Set chunk to empty so that we can call request_parser to get
         -- anything existing in the parser to flush.
         chunk = ''
```

bfredl commented:
Just did some quick testing with clangd. Clangd works good with unchanged buffer, but something goes wrong along the way with buffer updates (I think, clangd complains about "missing AST for current document" and stuff).

Commit:
Merge remote-tracking branch 'origin/master' into lsp-review

Commit:
Add support for vim.NIL

Commit:
Use vim.rpcnotify/rpcnotify

Commit:
Fix test for windows, hopefully.

Commit:
Update tests and fix some edge cases.

Commit:
Remove set_timeout

Commit:
Change tabs and clean up tests.

Commit:
Rename attach_to_buffer to buf_attach_client

Commit:
Disable incremental sync for now and improve tests.

- Add a timeout for tests. 1 second
- Wrap the test call in pcall

Commit:
Merge remote-tracking branch 'origin/master' into lsp-review

Commit:
Fix for Lua 5.1 test.

Commit:
Use vim.validate

norcalli commented:
Incremental sync has been discovered to have problems which will be addressed in a future PR. Full sync works for every client we've tested so far which otherwise supports incremental sync and is a decently safe choice.

I'm done changing this PR, now, save for any review concerns or comments. I would recommend making sure that the functions I have put available in `shared.lua` are acceptable, as that's something I'm unsure of. Also to make sure that the docs are acceptable.

clason commented on runtime/doc/lsp.txt:
```diff
@@ -0,0 +1,652 @@
+  These are functions which can be used to create a simple configuration which
+  will start a language server for a list of filetypes based on the |FileType|
+  event.
+  It will lazily start start the server, meaning that it will only start once
```
Typo: `start start`

norcalli commented:
In progress: diagnostics leveraging extmarks (thanks @bfredl) with the ability to show multiple diagnostics for a line/range:
![image](https://user-images.githubusercontent.com/296363/68689129-c2992780-0524-11ea-8883-e051153eecbd.png)
![image](https://user-images.githubusercontent.com/296363/68689174-cdec5300-0524-11ea-9f9c-9c7143943471.png)

mcepl commented:
One small regression from https://github.com/neovim/neovim/pull/10222 : there used to be function ``lsp#text_document_completion()`` which could be mapped to ``inoremap <buffer> <c-n> <c-r>=lsp#text_document_completion()<CR>``. I cannot find it now or any of its equivalent (but documentation mentions that there should be callback for ``textDocument/completion``).

norcalli commented:
New default diagnostics style ![image](https://user-images.githubusercontent.com/296363/68696722-59b8ac00-0532-11ea-8328-d838ccd84ad0.png)

@mcepl Indeed. I'm going to add that back in. I took it out for some (good?) reason, but it'll be easy to add back in.

clason commented:
tiny bike shed: replace the last (shown) indicator by `‣`?

pwntester commented:
Would be great to be able to align diagnostics messages to the right

bfredl commented:
This works great now in my testing with clangd.  As soon as diagnostics as above are available I will use it as my daily driver :]

I will look through the code (mainly the backend) once more, but I think we should be able to merge this in the following days. Then we have the 0.5 cycle to make it a complete experience with the best implementation of every specific method.

Commit:
Add publishDiagnostics with virtual text support.

Commit:
Fix :|

Commit:
Fix support for multiline diagnostics.

Commit:
Fix support for multiline diagnostics.

Commit:
Add location list support for diagnostics.

Commit:
Merge branch 'lsp-review' of github.com:norcalli/neovim into lsp-review

Commit:
Remove loclist by default.

FelschR commented:
I've tried setting this up with omnisharp-roslyn and getting some errors:
```
Error executing vim.schedule lua callback: ...lsp/share/nvim/runtime/lua/vim/lsp/builtin_callbacks.lua:253: attempt to index a nil value
Press ENTER or type command to continue
Error executing luv callback:
...-unwrapped-pr-lsp/share/nvim/runtime/lua/vim/lsp/rpc.lua:410: ...-unwrapped-pr-lsp/share/nvim/runtime/lua/vim/lsp/rpc.lua:93: invalid header line "Failed to launch '/bin/bash' with args, '-c 'uname -s -m''\
Content-Length: 239"
```

This is my lsp config:
```viml
call lsp#add_filetype_config({
  \ 'name': 'csharp',
  \ 'filetype': ['cs', 'csx'],
  \ 'cmd': ['omnisharp', '-lsp'],
  \ })
```

norcalli commented:
@FelschR based on your error, I'm guessing that the command is outputting some garbage to stdout before actually running. I've noticed this problem with some LSP servers. If you run `omnisharp -lsp 2>/dev/null` by itself, does it output stuff like that message about "/bin/bash" before going silent?

FelschR commented:
@norcalli It actually doesn't have any output.

norcalli commented:
@FelschR regardless, I think I know what the error is and I'll push a change to make the LSP client more permissive of invalid initial server output. If you'd like more help in realtime, I hang out in the gitter, btw.

Commit:
Add fix for servers which send garbage at the start.

norcalli commented:
@FelschR should be fixed.

Commit:
Cleaner assert handling in rpc code.

Commit:
A client can shut down before callbacks.

Commit:
Commit to rpc fix for bad servers.

Commit:
Refactor, bugfix, repeat.

ilAYAli commented on runtime/doc/lsp.txt:
```diff
@@ -0,0 +1,661 @@
+================================================================================
+                                                                       *lsp-api*
+
+Neovim exposes a API for the language server protocol. To get the real benefits
```
s/a API/an API/

bfredl commented on runtime/lua/vim/lsp/protocol.lua:
```diff
@@ -0,0 +1,936 @@
+
+
+--[=[
+-- Useful for interfacing with:
```
when is this used? to add new enums from spec revisions?

norcalli replied:
Yeah. I basically just do `gf` on the url and then copy stuff over and use the functions to transform it.

bfredl commented on runtime/lua/vim/lsp/protocol.lua:
```diff
@@ -0,0 +1,936 @@
+
+--[=[
+--Text document specific client capabilities.
+export interface TextDocumentClientCapabilities {
```
What is this? stuff we do not use now, but plan to implement eventually?

norcalli replied:
@bfredl Reference for things we should implement later. I can probably just delete it, but it was helpful to have around.

Commit:
Fix linter/docs.

bfredl commented on runtime/lua/vim/lsp/util.lua:
```diff
@@ -0,0 +1,557 @@
+      local highlight_name = "LspDiagnostics"..severity_name
+      local hi_info = highlights[severity] or default_color
+      -- Try to fill in the foreground color with a sane default.
+      if not hi_info.guifg and hi_info.guibg then
```
[TODO for after merge]. This seems oddly specific. Diagnostic code should just use `LspDiagnostics{severity}` which should be defined `:hi default` and thus overridable by user anyway. A mechanism for generating bright/dark colors should be separate from that, and invoked together by user if they want to.

norcalli replied:
Sure, I just had the code existing from my colorizer.

bfredl commented on runtime/lua/vim/lsp/util.lua:
```diff
@@ -0,0 +1,557 @@
+      for i = 1, #line_diags - 1 do
+        table.insert(virt_texts, {"■", severity_highlights[line_diags[i].severity]})
+      end
+      local last = line_diags[#line_diags]
```
is the spec guaranteeing that a diagnostic with highest severity comes last? (just wondering, we can improve this later)

norcalli replied:
Nope. Any delivery order, and I did consider the same thing. I'll slap a [TODO] here so we can Ctrl-F.

bfredl commented on runtime/lua/vim/lsp.lua:
```diff
@@ -0,0 +1,1054 @@
+      -- See https://github.com/palantir/python-language-server/commit/cfd6675bc10d5e8dbc50fc50f90e4a37b7178821#diff-f68667852a14e9f761f6ebf07ba02fc8 for an example of pyls handling both.
+      --]=]
+      elseif true or text_document_did_change == protocol.TextDocumentSyncKind.Full then
+        changes = full_changes(client)
```
TODO for later: we should throttle full buffer updates in some situation like global change.

bfredl commented on test/functional/plugin/lsp/lsp_spec.lua:
```diff
@@ -0,0 +1,634 @@
+    end)
+
+    -- TODO(askhan) we don't support full for now, so we can disable these tests.
+    -- it('should check the body and didChange incremental', function()
```
use `s/it/(pending(/` instead of comment out.

Commit:
Change incremental tests to pending.

bfredl commented on runtime/lua/vim/lsp/builtin_callbacks.lua:
```diff
@@ -0,0 +1,296 @@
+  end
+end
+
+local function update_tagstack()
```
TODO: this doesn't seem specific. Perhaps there should be a single-call API function to emulate tagstack?

norcalli replied:
That was a leftover from the last PR, so tbh I'm not exactly sure what it does at all.

justinmk replied:
Now that 'tagfunc' is available this might take the shape of https://github.com/neovim/neovim/pull/11280/files/0173bdf98be6f867d1b316d4d2ac87f7a93d95e4..c6afad78d39aa77a4d372759336018ef6e101dab

bfredl commented on runtime/lua/vim/lsp.lua:
```diff
@@ -0,0 +1,1055 @@
+    return
+  end
+  for client_id in pairs(client_ids) do
+    -- This is unlikely to happen. Could only potentially happen in a race
```
TODO: reconsider this. I think it can happen if `callback()` invokes event processing which closes the client.

bfredl commented on runtime/lua/vim/lsp.lua:
```diff
@@ -0,0 +1,1055 @@
+    end
+    local _ = log.debug() and log.debug(log_prefix, "client.request", client_id, method, params, callback)
+    -- TODO keep these checks or just let it go anyway?
+    if (not client.resolved_capabilities.hover and method == 'textDocument/hover')
```
TODO: we should make this more systematic, like a map from capability names to method names or something.

bfredl commented on runtime/autoload/lsp.vim:
```diff
@@ -0,0 +1,45 @@
+function! lsp#add_filetype_config(config) abort
```
TODO: reconsider this layer when we have `v:lua`.

norcalli commented:
For future travelers here, I'm trying to build an interface for common configurations so that we can have something similar to coc's servers except all in one place here https://github.com/norcalli/nvim-common-lsp. Help me get your favorite server to have an easy to use interface.

Also thanks to @h-michael for keeping the old PR alive, updating it, and answering my questions, and @tjdevries for the initial PR :) Without them, this wouldn't have been possible.

h-michael commented:
@nocalli When I got up in the morning, there was a present!  Thank you for the great achievement. 
@tjdevries This time came because you suggested, started the discussion, and started the implementation.
@neovim/core Thanks for adding and reviewing many important features needed for this.

tbodt commented:
I have an alternate idea for configuration:

```
lsp#add_config('clangd', {'cmd': 'clangd'})
autocmd FileType c,cpp lsp#set_server('clangd')
```

This has the huge (to me) advantage of allowing you to define project-specific language server configurations:

```
lsp#add_config('special clangd', {'cmd': ['~/project/prebuilt/bin/clangd', '-log=verbose']})
autocmd FileType c,cpp lsp#set_server('special clangd')
```

Does this sound like a good idea? I'd be happy to write a PR for it.

jrop commented:
@tbodt 
> project-specific

This would be pretty sweet, actually

norcalli commented:
@tbodt I'm guessing your comment is in response to the `add_filetype_config` interface, which, I agree, I was never a fan of. It felt too low value for something that a user could easily do themselves with just a little bit of Lua, but I guess the thought was to provide something that anyone could use. The interface that I prefer is the `start_client` and `buf_attach_client` pair. With those, you can do everything you want.

So, for instance, your suggestion, while better than the `add_filetype_config` interface, could easily be achieved with:

```lua
clangd_base_config = {
  cmd = 'clangd';
}
clangd_clients = {}

vim.api.nvim_command [[autocmd FileType c,cpp lua clangd_clients[vim.lsp.start_client(vim.tbl_extend("error", clangd_base_config, { root_dir = vim.fn.expand("%:h") }))] = true]]
```

Or something literally anything. The point is that the amount of code required is very minimal if you use Lua, whereas I found all the vimL interfaces difficult to design for because they are limited.

However, as to your point of per-project configuration, if you look at the docs at https://github.com/norcalli/nvim-common-lsp#gopls, you'll see how you can do that without needing something like .lvimrc. You use the `root_dir` function to signal what is or isn't a new project based on its root directory. Again, this isn't a vimL interface, but, tbh, I'm a Lua expert, not a vimL one.

So I think I would remove all the "utility" interfaces I've added and defer to the Lua interface `start_client/buf_attach_client` as the default and allow others/the community to decide what a good vimL interface should be.

tbodt commented:
Agreed that `start_client`/`attach_buf_client` can do all things. I'm not a big fan of Lua over VimL though, mainly because my entire vimrc is VimL and it feels weird to have two languages. I'll try and think of a good VimL interface for this. The root_dir interface from common-lsp is pretty sweet, maybe I'll come up with a VimL design based on that.

jonhoo commented:
:tada:

It would be really handy to have a "quick-start" guide for this feature. Just how to get set up with the basics working!

mcepl commented:
> tada
> 
> It would be really handy to have a "quick-start" guide for this feature. Just how to get set up with the basics working!

``:he lsp.txt`` has something, but yes this and other questions are being worked on #11389.

andymass commented:
Does this handle utf-16 properly?

clason commented:
@andymass not yet; there are still some issues with missing byte--char conversions, but it's on top of the list mentioned in the issue mcepl mentioned.

norcalli commented:
@andymass yeah, that's first on my agenda for today. It should be a transparent fix which makes any existing callbacks just work.
